src/index.ts

```diff
@@ -0,0 +1,221 @@
+  }
+  return options;
+}
+
+
+export interface ProjectOptions {
+  /**
+   * Path to the root of the project on the filesystem. This can be an absolute
+   * path, or a path relative to the current working directory. Defaults to the
+   * current working directory of the process.
+   */
+  root?: string;
+
+  /**
+   * The path relative to `root` of the entrypoint file that will be served for
+   * app-shell style projects. Usually this is index.html.
+   */
+  entrypoint?: string;
+
+  /**
+   * The path relative to `root` of the app shell element.
+   */
+  shell?: string;
+
+  /**
+   * The path relative to `root` of the lazily loaded fragments. Usually the
+   * pages of an app or other bundles of on-demand resources.
+   */
+  fragments?: string[];
+
+  /**
+   * List of glob patterns, relative to root, of this project's sources to read
+   * from the file system.
+   */
+  sources?: string[];
+
+  /**
+   * List of file paths, relative to the project directory, that should be included
+   * as dependencies in the build target.
+   */
+  dependencies?: string[];
+}
+
+export class ProjectConfig {
+
+  readonly root: string;
+  readonly entrypoint: string;
+  readonly shell: string;
+  readonly fragments: string[];
+  readonly sources: string[];
+  readonly dependencies: string[];
+
+  readonly allFragments: string[];
+
+  /**
+   * Given an absolute file path to a polymer.json-like ProjectOptions object,
+   * read that file. If no file exists, null is returned. If the file exists
+   * but there is a problem reading or parsing it, throw an exception.
+   */
+  static loadOptionsFromFile(filepath: string): ProjectOptions {
+    let configParsed: ProjectOptions;
+    try {
+      const configContent = fs.readFileSync(filepath, 'utf-8');
+      configParsed = JSON.parse(configContent);
+    } catch (error) {
+      // swallow "not found" errors because they are so common / expected
+      if (error.code === 'ENOENT') {
+        logger.debug('no polymer config file found', {file: filepath});
+        return null;
+      }
+      // otherwise, throw an exception
+      throw error;
+    }
+
+    return configParsed;
+  }
+
+  /**
+   * Given an absolute file path to a polymer.json-like ProjectOptions object,
+   * return a new ProjectConfig instance created with those options.
+   */
+  static loadConfigFromFile(filepath: string): ProjectConfig {
+    let configParsed = ProjectConfig.loadOptionsFromFile(filepath);
+    if (!configParsed) {
+      return null;
+    }
+    return new ProjectConfig(configParsed);
+  }
+
+  /**
+   * constructor - given a ProjectOptions object, create the correct project
+   * configuration for those options. This involves setting the correct
+   * defaults, validating options, warning on deprecated options, and
+   * calculating some additional properties.
+   */
+  constructor(options: ProjectOptions) {
+    options = options || {};
+    options = fixDeprecatedOptions(options);
+
+    /**
+     * root
```

Comment:
These comments are more useful to separate the different sections of the constructor. From original PR: 

> I wrote the constructor this way because there are some ordering concerns with setting these properties that are hidden and/or much harder to follow when refactored into different helper functions.